core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -172,6 +184,44 @@ public Result applyFailedShards(ClusterState clusterState, List<FailedRerouteAll
         return buildResultAndLogHealthChange(allocation, "shards failed [" + failedShardsAsString + "] ...");
     }
 
+    private ClusterState removeStaleAllocationIds(ClusterState clusterState, List<FailedRerouteAllocation.StaleShard> staleShards) {
+        MetaData oldMetaData = clusterState.metaData();
+        MetaData.Builder metaDataBuilder = null;
+        for (Map.Entry<Index, List<FailedRerouteAllocation.StaleShard>> entry : staleShards.stream().collect(
+            Collectors.groupingBy(fs -> fs.shardId.getIndex())).entrySet()) {
+            final IndexMetaData oldIndexMetaData = oldMetaData.getIndexSafe(entry.getKey());
+            IndexMetaData.Builder indexMetaDataBuilder = null;
+            for (FailedRerouteAllocation.StaleShard staleShard : entry.getValue()) {
+                // remove allocation ids for stale shards
+                int shardNumber = staleShard.shardId.getId();
+                Set<String> inSyncAllocations = indexMetaDataBuilder == null ?
+                    oldIndexMetaData.inSyncAllocationIds(shardNumber) :
+                    indexMetaDataBuilder.getInSyncAllocationIds(shardNumber);
+                String shardAllocationId = staleShard.allocationId;
+                if (inSyncAllocations.contains(shardAllocationId) && inSyncAllocations.size() > 1) {
```

Comment:
I'll rework this part. The reason why `inSyncAllocations.size() > 1` is a safeguard so that we don't end up with an empty set of `inSyncAllocations` in case of some bugs. An empty set would mean that only stale copies exist and a primary cannot be allocated anymore (which should never be the case). I will add an assertion to check we never run into that situation.